### ISSUE TITLE
Handle rdkafka Message consumption error: Fatal (Local: Fatal error)

### ISSUE THREAD

**Issue by fpytloun** [vectordotdev/vector#13665]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

When kafka has under-replicated partitions, Vector crashes after a while.

With disabled rack awareness, error looks like this:
```
2022-07-21T16:47:10.352136Z ERROR source{component_kind="source" component_id=in_kafka_access component_type=kafka component_name=in_kafka_access}: vector::internal_events::kafka: Failed to read message. error=Message consumption error: Fatal (Local: Fatal error) error_code="reading_message" error_type="reader_failed" stage="receiving"
thread 'vector-worker' panicked at 'called `Option::unwrap()` on a `None` value', /cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-0.27.0/src/topic_partition_list.rs:201:43
```

https://github.com/fede1024/rust-rdkafka/blob/master/src/topic_partition_list.rs#L201

With rack awareness enabled:

```
*** /cargo/registry/src/github.com-1ecc6299db9ec823/rdkafka-sys-4.0.0+1.6.1/librdkafka/src/rdkafka_partition.c:346:rd_kafka_toppar_set_fetch_state: assert: thrd_is_current(rktp->rktp_rkt->rkt_rk->rk_thread) ***
```

which is probably this issue: https://github.com/edenhill/librdkafka/issues/3569 (vector: https://github.com/vectordotdev/vector/issues/8750)

For first issue, Vector should probably handle this kind of errors and re-initialize client without panicking.

### Configuration

```text
## Kafka source
    [sources.in_kafka_access]
      type = "kafka"
      bootstrap_servers = "kafka-0.sre-mydomain.pa4-par-gc.compass.tenant.int.ves.io:9093,kafka-1.sre-ves.pa4-par-gc.compass.tenant.int.ves.io:9093,kafka-2.sre-ves.pa4-par-gc.compass.tenant.int.ves.io:9093"
      group_id = "vector_access"
      topics = ["^fluentd.svcfw.(apiaccess|secevent)(_.*)?$"]
      librdkafka_options."client.id" = "${HOSTNAME}.pa4-par-gc-int-mydomain-io"
      librdkafka_options."client.rack" = "pa4-par-gc-int-mydomain-io"
      librdkafka_options."group.instance.id" = "${HOSTNAME}.pa4-par-gc-int-mydomain-io"
      # Prefer roundrobin balancing to spread load more evenly
      librdkafka_options."partition.assignment.strategy" = "roundrobin,range"
      librdkafka_options."session.timeout.ms" = "60000"
      topic_key = "_topic"
      partition_key = "_partition"
      offset_key = "_offset"

      tls.enabled = true
      tls.ca_file = "/domain/secrets/identity/server_ca_with_compass.crt"
      tls.crt_file = "/domain/secrets/identity/client.crt"
      tls.key_file = "/domain/secrets/identity/client.key"
```

### Version

vector 0.21.1 (x86_64-unknown-linux-gnu)

### References

https://github.com/fede1024/rust-rdkafka/issues/279

**Comment by jszwedko** [comment]
Hi @fpytloun ! Thanks for this report.

I think this will be fixed by #10302 since the issue was fixed in rdkafka in 0.28.0 (https://github.com/fede1024/rust-rdkafka/blob/master/changelog.md).

**Comment by fpytloun** [comment]
FYI: this issue also occurs on producer. When there's some glitch or issue with upstream kafka, I start to see this message:

```
2022-08-03T06:52:00.031666Z ERROR sink{component_kind="sink" component_id=out_fluent_kafka component_type=kafka component_name=out_fluent_kafka}:request{request_id=3794715}: vector_core::stream::driver: Service call failed. error=KafkaError (Message production error: Fatal (Local: Fatal error)) request_id=3794715
```

And see no output into Kafka until I restart vector :-o
